src/zinit/mod.rs

```
@@ -168,39 +212,117 @@ impl ZInit {
             }),
         };
 
-        let service = service.read().await.clone();
+        let service = service.read().await.status();
         Ok(service)
     }
-
-    pub async fn shutdown(&self) -> Result<()> {
-        info!("shutting down");
-        *self.shutdown.write().await = true;
-        loop {
-            let table = self.services.read().await;
-            let mut to_kill: Vec<String> = Vec::new();
-            for (name, service) in table.iter() {
-                let service = service.read().await;
-                if service.state == State::Running || service.state == State::Spawned {
-                    info!("service '{}' is scheduled for a shutdown", name);
-                    to_kill.push(name.into());
+    async fn wait_kill(
+        name: String,
+        ch: mpsc::Sender<String>,
+        mut rx: watch::Receiver<State>,
+        shutdown_timeout: u64,
+    ) -> Result<()> {
+        debug!("kill_wait {}", name);
+        let cp = name.clone();
+        let fut = timeout(
+            std::time::Duration::from_secs(shutdown_timeout),
+            async move {
+                while rx.changed().await.is_ok() {
+                    let new_state = rx.borrow();
+                    debug!(
+                        "received a state change notification for service {} to {:?}",
+                        cp, new_state
+                    );
+                    if *new_state != State::Running && *new_state != State::Spawned {
+                        return;
+                    }
+                }
+            },
+        );
+        let _ = fut.await;
```

You need to know here if the service has actually exited, or timedout. If the service has timed out you need to `KILL -9` before returing (don't wait for it's state to change, a SIGKILL is granted to end the process) We don't want to leave hanging processes inside a container that doesn't die because they are not handing sigterm

---

Ignoring the error was done considering that the goal of the shutdown phase is to notify the services to perform graceful shutdown, not to cleanup the process tree. This is also why only the services are shutdown not the processes (the services children gets reassigned to `zinit` process if the service didn't clean them up). And whoever started zinit as the `init` process is responsible for cleaning up the process tree. For example, `docker` waits for some time (10 seconds by default) and force kills `init` if it didn't return, and the `kernel` sends a SIGKILL to all process in the container whenever the `init` process dies, regardless of its termination being graceful or forceful.

The benefit of not SIGKILLing is that we give the service a chance to continue cleaning up while other services are not done. And in case the hanging service depends on its children being alive for its cleanup, leaving it hanging will make the hanging service probably fail (instead of definitely fail).